File: tests/cross-language/verify_token.py
#!/usr/bin/env python3
"""
Cross-language token verification test.

Verifies that a Cloak token minted by Rust can be decoded and verified in Python.
Token format: base64url(claims_json).hex(hmac_sha256)
HMAC is computed over the base64url payload string (not the raw JSON).

Usage: python verify_token.py [fixture.json]
"""

import base64
import hashlib
import hmac
import json
import sys


def verify_token(token: str, key_hex: str) -> dict:
    """Verify a Cloak token and return decoded claims."""
    key = bytes.fromhex(key_hex)

    # Split on last '.' to separate payload from signature
    dot_idx = token.rfind(".")
    if dot_idx == -1:
        raise ValueError("Malformed token: no separator")

    payload_b64 = token[:dot_idx]
    signature_hex = token[dot_idx + 1 :]

    # Compute expected HMAC-SHA256 over the base64url payload string
    expected = hmac.new(key, payload_b64.encode("utf-8"), hashlib.sha256).hexdigest()

    # Constant-time comparison
    if not hmac.compare_digest(expected, signature_hex):
        raise ValueError("Invalid signature")

    # Decode payload (base64url, no padding)
    # Add padding back
    padding = 4 - len(payload_b64) % 4
    if padding != 4:
        payload_b64 += "=" * padding

    claims_json = base64.urlsafe_b64decode(payload_b64)
    return json.loads(claims_json)


def main():
    fixture_path = sys.argv[1] if len(sys.argv) > 1 else "tests/cross-language/fixture.json"

    with open(fixture_path) as f:
        fixture = json.load(f)

    token = fixture["token"]
    key_hex = fixture["key_hex"]
    expected_claims = fixture["claims"]

    # Verify
    claims = verify_token(token, key_hex)

    # Check fields
    assert claims["job_id"] == expected_claims["job_id"], (
        f"job_id mismatch: {claims['job_id']} != {expected_claims['job_id']}"
    )
    assert claims["agent_class"] == expected_claims["agent_class"], (
        f"agent_class mismatch"
    )
    assert len(claims["services"]) == 1
    assert claims["services"][0]["service"] == "episteme"

    print(f"PASS: Python verified token for job_id={claims['job_id']}")
    print(f"  Claims: {json.dumps(claims, indent=2)}")


if __name__ == "__main__":
    main()

File: tests/cross-language/verify_token.ts
#!/usr/bin/env -S npx tsx
/**
 * Cross-language token verification test.
 *
 * Verifies that a Cloak token minted by Rust can be decoded and verified in TypeScript.
 * Token format: base64url(claims_json).hex(hmac_sha256)
 * HMAC is computed over the base64url payload string (not the raw JSON).
 *
 * Usage: npx tsx verify_token.ts [fixture.json]
 *
 * No dependencies needed — uses Node.js built-in crypto module.
 */

import { createHmac, timingSafeEqual } from "crypto";
import { readFileSync } from "fs";

function verifyToken(
  token: string,
  keyHex: string
): Record<string, unknown> {
  const key = Buffer.from(keyHex, "hex");

  // Split on last '.' to separate payload from signature
  const dotIdx = token.lastIndexOf(".");
  if (dotIdx === -1) throw new Error("Malformed token: no separator");

  const payloadB64 = token.slice(0, dotIdx);
  const signatureHex = token.slice(dotIdx + 1);

  // Compute expected HMAC-SHA256 over the base64url payload string
  const expected = createHmac("sha256", key)
    .update(payloadB64, "utf-8")
    .digest("hex");

  // Constant-time comparison
  const expectedBuf = Buffer.from(expected, "utf-8");
  const actualBuf = Buffer.from(signatureHex, "utf-8");
  if (
    expectedBuf.length !== actualBuf.length ||
    !timingSafeEqual(expectedBuf, actualBuf)
  ) {
    throw new Error("Invalid signature");
  }

  // Decode base64url payload
  // Node's Buffer handles base64url natively with 'base64url' encoding
  const claimsJson = Buffer.from(payloadB64, "base64url").toString("utf-8");
  return JSON.parse(claimsJson);
}

function main() {
  const fixturePath = process.argv[2] || "tests/cross-language/fixture.json";
  const fixture = JSON.parse(readFileSync(fixturePath, "utf-8"));

  const { token, key_hex: keyHex, claims: expectedClaims } = fixture;

  const claims = verifyToken(token, keyHex);

  // Assertions
  if (claims.job_id !== expectedClaims.job_id) {
    throw new Error(
      `job_id mismatch: ${claims.job_id} != ${expectedClaims.job_id}`
    );
  }
  if (claims.agent_class !== expectedClaims.agent_class) {
    throw new Error("agent_class mismatch");
  }
  const services = claims.services as Array<Record<string, unknown>>;
  if (services.length !== 1 || services[0].service !== "episteme") {
    throw new Error("services mismatch");
  }

  console.log(`PASS: TypeScript verified token for job_id=${claims.job_id}`);
  console.log(`  Claims: ${JSON.stringify(claims, null, 2)}`);
}

main();
